### PR TITLE
Fix inconsistent Link UI in Nav block list view editor

### LINKED ISSUE #50601: Link Control UI Opens for Last Inserted Block when returning to the Site Editor Right Sidebar Navigation Inserter
Link Control opens when returning to side inserter component

### The Problem

It's unclear (without reading the code), why that Link UI would open anytime the component is mounted. It's due to[ a `useEffect` hook that opens the UI](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/off-canvas-editor/block-contents.js#L79-L92) if:
- There is a `lastInsertedBlockId`
- The last inserted block is allowed to be inside the navigation block
- The block does not have a URL set

These things are all still true when the component rerenders, so the Link UI shows again when it's mounted.

The Link UI also always shows when the first block in the navigation is missing a url, as the `lastInsertedBlockId` returns the first block in the array on load. I believe this is because on load it considers all the existing blocks as the last inserted blocks.

### How to Solve it
I believe the intention is that **the LinkUI shows ONLY once when the link is added**.  If so, we can set a state to see if that UI has already shown or not, or set the useEffect as such that it only runs if the `lastInsertedBlockId` has changed.

### Test Reproduction Steps
- In the Site Editor Right Sidebar Navigation Inserter:
- Click the + at the bottom to add a new navigation item
- It opens the Link Control UI
- Click Cancel (do not set a link)
- Click to another item or anywhere else to remove the Right Sidebar Navigation Inserter from the view
- Return to the Sidebar Navigation
- The Link UI Will Open

### PR DESCRIPTION
**Pull request by getdave**
## What?

Fixes inconsistent and unexpected appearance of Link UI within the Nav block list view editing. Also ensures the correct value appears in the UI when multiple Nav blocks that reference the _same menu_ are in the template.

Also removes code that is no longer used.

Fixes https://github.com/WordPress/gutenberg/issues/50601 and https://github.com/WordPress/gutenberg/issues/50733

## Why?

There's a [long explanation of the problem(s) in the Issue](https://github.com/WordPress/gutenberg/issues/50601#issuecomment-1552964876).

### TL;DR version

- The Nav block list view editing needs to know about the last block that was inserted in order to display the Link UI for that block.
- On `trunk` this relies on global state for `lastInsertedBlock` to find "the last block inserted into the editor" which seems logical, but in reality is not what we want to do.
- Instead we actually only care about "the last block that was _just_ added _into the list view_ by the user". The distinction is subtle but critical.
- Currently on `trunk` the Link UI will appear unexpectedly on "re-mount" if the last inserted block doesn't have a `url`. Moreover, if there are multiple Nav blocks in the template referencing the same menu, then the Link UI will appear but with the "wrong" (unexpected) value - this value comes from another block on the page which is "syncing" with the menu being altered in the first blocks list view.
- If the above is all very confusing please see the Issues below and experience the problems for yourself. I think this is an important step of being able to review this PR:
    - https://github.com/WordPress/gutenberg/issues/50601
    - https://github.com/WordPress/gutenberg/issues/50733

## How?

Moves all state about the "last inserted block" to rely on state that's local to the current list view. This is achieved by moving the getter/setter for "last inserted" within the List View and putting it within the `Context` object of List View.

As a result of utilising context, we are then able to pass the state about the "last inserted block" to the render prop callback which determines whether to display the Link UI.

Within the code in the Nav block which provides the render prop to the List View that is used in it's "list view editing mode", we then remove all global state and move to using the state provided to the render callback to determine whether to show the Link UI and also which block attributes should be shown in the Link UI.

The result of all this is that state about "last inserted" is made entirely local to the current List View instance and thus we avoid unexpected interactions that can come from consuming global editor state about "last inserted".

## Testing Instructions

Test against the scenarios and testing instructions in both Issues - they should both no longer be in evidence:
    - https://github.com/WordPress/gutenberg/issues/50601
    - https://github.com/WordPress/gutenberg/issues/50733

Also please:

- test the Link UI still shows when adding Submenus via the options menu in the List View.
- Stress test against other common scenarios. 

Our e2e tests for offcanvas provide some coverage but human trial/error is also valuable.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/444434/9143e79e-be10-4fec-9e1b-0ed546f10edf

### CODE REVIEW

**Comment by getdave** [comment]
## Todo

Try the same technique for `BlockSetitingsMenu` - let's hope this works 🙏

**Comment by getdave** [comment]
## Todo 

Prior to merging we need to add test coverage for the scenarios we found in the two related Issues.

**Comment by getdave** [comment]
Pinging @andrewserong as I know he's been working a lot on list view. This PR fixes the issues but curious how you feel about (partially) exposing an API (`setInsertedBlock`) which allows setting of state within the List View from outside.

This is a form of IoC but I wanted to get some more opinions before this ends up landing.

**Update**: @scruffian's makes a good point that this API is only exposed via a private mechanic and even then only to a specific render prop.

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 1]
```diff
@@ -572,11 +588,17 @@ test.describe( 'Navigation block', () => {
 			// Expect to see the Link creation UI be focused.
 			const linkUIInput = linkControl.getSearchInput();
 
+			// Coverage for bug whereby Link UI input would be incorrectly prepopulated.
+			// It should:
+			// - focused - should not be in "preview" mode but rather ready to accept input.
```
Comment (scruffian) [inline test/e2e/specs/editor/blocks/navigation.spec.js]:
```suggestion
			// - be focused - should not be in "preview" mode but rather ready to accept input.
```

File: test/e2e/specs/editor/blocks/navigation.spec.js [hunk 2]
```diff
@@ -821,6 +843,92 @@ test.describe( 'Navigation block', () => {
...
 		} );
+
+		test( `does not display link interface for blocks that have not just been inserted`, async ( {
+			admin,
+			page,
+			editor,
+			requestUtils,
+			linkControl,
+		} ) => {
+			// Provides coverage bug whereby the Link UI would be unexpectedly displayed for the last
```
Comment (scruffian) [inline test/e2e/specs/editor/blocks/navigation.spec.js]:
```suggestion
			// Provides coverage for a bug whereby the Link UI would be unexpectedly displayed for the last
```

File: packages/block-editor/src/components/list-view/index.js [hunk 3]
```diff
@@ -221,6 +227,8 @@ function ListViewComponent(
 			BlockSettingsMenu,
 			instanceId,
 			renderAdditionalBlockUI,
+			insertedBlock,
+			setInsertedBlock,
```
Comment (scruffian) [inline packages/block-editor/src/components/list-view/index.js:231]: We are stuffing even more stuff into context. Is that a problem?

File: packages/block-library/src/navigation/edit/leaf-more-menu.js [hunk 4]
```diff
@@ -24,7 +24,13 @@ const BLOCKS_THAT_CAN_BE_CONVERTED_TO_SUBMENU = [
 	'core/navigation-submenu',
 ];
 
-function AddSubmenuItem( { block, onClose, expandedState, expand } ) {
+function AddSubmenuItem( {
```
Comment (scruffian) [inline packages/block-library/src/navigation/edit/leaf-more-menu.js:27]: A copy of this code is also in the Navigation section of Browse Mode. Since we don't show the Link UI there I don't think we need to make these changes there, I'm just noting it in case I've missed anything.

**Review by scruffian — APPROVED** [review]
This is looking good to me. From what I can see the setter is only exposed to renderAdditionalBlockUI , which is a private prop so I'm not concerned about that.

I think we added the concept of `lastInsertedBlockClientId` to the store for this PR. Should we also consider removing that?

File: packages/block-editor/src/components/list-view/index.js (hunk 3 above)
Comment (getdave) [reply packages/block-editor/src/components/list-view/index.js:231]: It's always a potential problem for performance because if the context changes and components which consume it (or their children) will be re-rendered.

The setter is stable, but the `insertedBlock` will change. 

However, pre-optimising context is challenging. Better to include in the `memo` and then optimise if/when List View becomes a problem.

File: packages/block-library/src/navigation/edit/leaf-more-menu.js (hunk 4 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/leaf-more-menu.js:27]: Good spot but it's going to be removed in https://github.com/WordPress/gutenberg/pull/50788 so no need to worry.

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js [hunk 5]
```diff
@@ -109,23 +80,45 @@ const MainContent = ( {
...
+		const blockHasExistingLinkValue = !! lastInsertedBlock?.url;
+		const blockSupportsLinkUI = BLOCKS_WITH_LINK_UI_SUPPORT?.includes(
+			lastInsertedBlock?.name
+		);
+		const currentBlockWasJustInserted =
+			lastInsertedBlock?.clientId === currentBlock.clientId;
+
+		const shouldShowLinkUIForBlock =
+			blockSupportsLinkUI &&
+			! blockHasExistingLinkValue && // don't re-show the Link UI if the block already has a link value.
```
Comment (jeryj) [inline packages/block-library/src/navigation/edit/menu-inspector-controls.js]: Is this check necessary? If it only appears after a user has inserted the block, it should never have a link, right? If so, I think we can remove all the checks for if there's a link or not.

**Review by jeryj — APPROVED** [review]
It's working well and fixes the bug! I much prefer that this explicitly sets the condition to show/hide rather than relying on a side-effect that happens to work (not showing the LinkControl on mount).

File: packages/block-library/src/navigation/edit/menu-inspector-controls.js (hunk 5 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/menu-inspector-controls.js]: Yeh you could be right there. Have you got time to test that and commit as it's nearly my EOD?

**Review by ajlende — APPROVED** [review]
✅ https://github.com/WordPress/gutenberg/issues/50601 is fixed
✅ https://github.com/WordPress/gutenberg/issues/50733 is fixed
✅ Code looks good to me
✅ New tests look correct and pass

I love to see it when we can find the root cause of something and close multiple issues with one PR. Good job!

File: packages/block-editor/src/components/list-view/block-contents.js [hunk 6]
```diff
@@ -66,7 +67,12 @@ const ListViewBlockContents = forwardRef(
 
 		return (
 			<>
-				{ renderAdditionalBlockUI && renderAdditionalBlockUI( block ) }
+				{ renderAdditionalBlockUI &&
+					renderAdditionalBlockUI(
+						block,
+						insertedBlock,
+						setInsertedBlock
```
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block-contents.js:74]: This now forms part of the API of `renderAdditionalBlockUI` as positional params. If and when `renderAdditionalBlockUI` becomes part of the stable API, `insertedBlock` and `setInsertedBlock` might not be applicable for some use cases, but we also might want to add additional params at some stage. Would it be worth switching `renderAdditionalBlockUI` to use an object as its param?

**Review by andrewserong — COMMENTED** [review]
Thanks for the ping!

The idea of ensuring that both the appender and the additional UI both have access to the inserted block / setInsertedBlock state sounds good to me. And like @scruffian mentioned, if it's a private API where it's being exposed, taking an iterative approach to this also sounds good 👍

One question I have from looking this over is about the `renderAdditionalBlockUI` function, and its positional params. Will we expect to add more params to this function in the future, and is it always expected that `renderAdditionalBlockUI` will return a React component? If so, I'm wondering if instead of it having a function name, if we could treat is as a component, and then instead of using positional params, it would receive props. So, something like:

```jsx
<AdditionalBlockUI
  block={ block }
  insertedBlock={ insertedBlock }
  setInsertedBlock={ setInsertedBlock }
/>
```

Apologies if I've missed some details on the benefits of using a function / callback approach instead!

To put the question another way: do we have an idea yet of what the final / stable shape of this API might look like? I know much of the recent efforts have been about removing the experimental off canvas editor component, which has been great to see. Now that that's happened, I'm curious if we can hone in on what a stable API shape might look like. There are a lot of moving parts in the ListView component, and it'd be good if we can try to stabilise and document some of these more nuanced areas if we can — if only so that we don't accidentally break anything in follow-up PRs.

I'll just CC @talldan, too, for visibility, too.

Nice work fixing up the inconsistencies! +1 to @ajlende's comments that it's great when fixing the root cause for something can close out multiple issues 🙂

File: packages/block-editor/src/components/list-view/block-contents.js (hunk 6 above)
Comment (getdave) [reply packages/block-editor/src/components/list-view/block-contents.js:74]: Thanks for the feedback on this. When I implemented this I also considered that but (for a reasons I can't recall) neglected to include it in the PR 😕 

I think it's a good idea. The useful state that might need to be made available to this render prop in future might be many things so we should convert to object.

As for using a component - that seems logical to me. This PR didn't implement that but it's definitely worth calling out. It's very similar to the `renderAppender` arg of `InnerBlocks` so we could follow the pattern set down there.

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block-contents.js:74]: Nice. Thanks for opening up the follow-up [issue](https://github.com/WordPress/gutenberg/issues/50827)!
